### PR TITLE
Feature 25 routes tasks drag and drop

### LINKED ISSUE #25: Routes tasks drag and drop
As a user I'd like to be able to drag and drop tasks into routes.

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #25

### CODE REVIEW

File: src/components/drive-planning/routes/unallocated-tasks-table.tsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+const UnallocatedTasksTable = ({ sites }: Props) => {
+  const { t } = useTranslation();
+  const { freightsApi, freightUnitsApi } = useApi();
+
+  const navigate = useNavigate({ from: "/drive-planning/routes" });
+  const [rowDragHandles, setRowDragHandles] = useState<UnallocatedTasksRowDragHandles>({});
+
+  const tasksQuery = useTasks({ assignedToRoute: false });
+
+  const getDistinctFreights = () => [...new Set((tasksQuery.data?.tasks ?? []).map((task) => task.freightId))];
```
Comment (lauriLukkarinen) [inline src/components/drive-planning/routes/unallocated-tasks-table.tsx:30]: This only compares object references, so it does not actually get distinct tasks, as they will have different refs when parsed from the API response.

File: src/components/drive-planning/routes/unallocated-tasks-table.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      freightsQueries.data?.map((freight) => ({
+        queryKey: [QUERY_KEYS.FREIGHT_UNITS_BY_FREIGHT, freight.id],
+        queryFn: () => freightUnitsApi.listFreightUnits({ freightId: freight.id }),
+      })) ?? [],
+    combine: (results) => ({
+      data: results.flatMap((result) => result.data).filter(DataValidation.validateValueIsNotUndefinedNorNull),
+    }),
+  });
+
+  const columns: GridColDef[] = useMemo(
```
Comment (lauriLukkarinen) [inline src/components/drive-planning/routes/unallocated-tasks-table.tsx:54]: Add row model type here
`GridColDef<T>[]`
to get proper types in renderCell functions

File: src/routes/drive-planning.routes.tsx [hunk 3]
```diff
@@ -36,16 +53,57 @@ export const Route = createFileRoute("/drive-planning/routes")({
...
   const sitesQuery = useSites();
 
-  const { date: selectedDate } = Route.useSearch();
-
   const [unallocatedDrawerOpen, setUnallocatedDrawerOpen] = useState(true);
+  const [activeDraggable, setActiveDraggable] = useState<Active | null>(null);
+  const [paginationModel, setPaginationModel] = useState<GridPaginationModel>({ page: 0, pageSize: 25 });
+  const [localTasks, setLocalTasks] = useState<TRouteTasks>({});
+
+  const localTasksBeforeDrag = useRef<null | TRouteTasks>(null);
```
Comment (lauriLukkarinen) [inline src/routes/drive-planning.routes.tsx:68]:
```suggestion
  const localTasksBeforeDrag = useRef<TRouteTasks | null>(null);
```

File: src/routes/drive-planning.routes.tsx [hunk 4]
```diff
@@ -68,6 +126,19 @@ function DrivePlanningRoutes() {
     onError: () => toast.error(t("drivePlanning.routes.errorToast")),
   });
 
+  const updateTask = useMutation({
+    mutationFn: (task: Task) => {
+      if (!task.id) return Promise.reject();
+      return tasksApi.updateTask({ taskId: task.id, task });
+    },
+    onSuccess: () => {
+      toast.success("Tehtävän siirto onnistui!");
```
Comment (lauriLukkarinen) [inline src/routes/drive-planning.routes.tsx:135]: Localize

File: src/types/index.ts [hunk 5]
```diff
@@ -11,6 +13,15 @@ export type NavigationItem = readonly [
   typeof SvgIcon | undefined,
 ];
 
+type UnallocatedTasksRowDragHandle = {
+  setActivatorNodeRef: (element: HTMLElement | null) => void;
+  attributes: DraggableAttributes;
+  listeners: SyntheticListenerMap | undefined;
+
```
Comment (lauriLukkarinen) [inline src/types/index.ts:20]:
```suggestion
```

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

File: src/components/drive-planning/routes/draggable-routes-tasks-table-row.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      data: draggableData,
+    });
+
+  const outline = useMemo(() => {
+    if (isDragging || active?.id.toString().includes(groupedTasksKey)) {
+      return `2px solid ${theme.palette.primary.main}`;
+    }
+  }, [isDragging, active, groupedTasksKey]);
+
+  const style: React.CSSProperties = useMemo(
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/drive-planning/routes/draggable-routes-tasks-table-row.tsx:42]: Resolved

File: src/components/drive-planning/routes/draggable-unallocated-tasks-table-row.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import { useTranslation } from "react-i18next";
+import { DraggableType, UnallocatedTasksRowDragHandles } from "../../../types";
+import { useNavigate } from "@tanstack/react-router";
+
+const DraggableUnallocatedTasksTableRow = ({
+  setRowDragHandles,
+  ...params
+}: GridRowProps & {
+  setRowDragHandles: Dispatch<SetStateAction<UnallocatedTasksRowDragHandles>>;
+}) => {
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/drive-planning/routes/draggable-unallocated-tasks-table-row.tsx:15]: Resolved

File: src/components/drive-planning/routes/expandable-routes-table-row.tsx [hunk 8]
```diff
@@ -1,52 +1,69 @@
...
+    data: { routeId: routeId },
   });
 
+  const droppableStyle = useMemo(
+    () => ({
+      outline: isOver ? "2px solid #4E8A9C" : "none",
+      outlineOffset: "-2px",
+    }),
+    [isOver],
+  );
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/drive-planning/routes/expandable-routes-table-row.tsx:55]: Resolved

File: src/components/drive-planning/routes/draggable-unallocated-tasks-table-row.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        [params.rowId]: undefined,
+      }));
+  }, [attributes, params.rowId, setRowDragHandles, listeners, setActivatorNodeRef]);
+
+  return (
+    <Tooltip title={t("drivePlanning.routes.unallocatedTasksTable.taskRowTooltip")} followCursor>
+      <div ref={setNodeRef}>
+        <GridRow
+          {...params}
+          onClick={() => navigate({ search: { freightId: params.row?.freightId, date: undefined } })}
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/drive-planning/routes/draggable-unallocated-tasks-table-row.tsx:54]: Resolved

File: src/components/drive-planning/routes/routes-table.tsx [hunk 10]
```diff
@@ -1,66 +1,62 @@
...
+  onUpdateRoute,
+}: Props) => {
   const { t } = useTranslation();
   const { tasksApi } = useApi();
-  const { cellModesModel, handleCellClick, handleCellModelsChange } = useSingleClickCellEditMode();
 
-  const [paginationModel, setPaginationModel] = useState<GridPaginationModel>({ page: 0, pageSize: 25 });
+  const onTasksCellClick = useCallback(({ field, row }: GridCellParams) => {
+    if (field !== "tasks") return;
+    if (!row.id) return;
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/drive-planning/routes/routes-table.tsx:45]: Resolved

File: src/components/drive-planning/routes/routes-tasks-table-row.tsx [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    }: Props,
+    ref: ForwardedRef<HTMLTableRowElement>,
+  ) => {
+    const { tasksApi } = useApi();
+    const { t } = useTranslation();
+    const queryClient = useQueryClient();
+    const navigate = useNavigate({ from: "/drive-planning/routes" });
+    const freightsQuery = useFreights();
+
+    const [menuCoordinates, setMenuCoordinates] = useState<{ clientX: number; clientY: number } | undefined>();
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/drive-planning/routes/routes-tasks-table-row.tsx:53]: Resolved

File: src/components/drive-planning/routes/routes-tasks-table-row.tsx [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+      return tasks.map((task) => {
+        if (!task.freightId) return null;
+        const foundFreight = freightsQuery.data?.freights.find((freight) => freight.id === task.freightId);
+
+        if (!foundFreight?.id) return null;
+
+        return (
+          <ListItemButton
+            key={task.id}
+            onClick={() => navigate({ search: { freightId: foundFreight.id, date: undefined } })}
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/drive-planning/routes/routes-tasks-table-row.tsx:67]: Resolved

File: src/components/drive-planning/routes/unallocated-tasks-table.tsx (hunk 2 above)
Comment (villejuutila) [reply src/components/drive-planning/routes/unallocated-tasks-table.tsx:54]: Resolved

File: src/routes/drive-planning.routes.tsx (hunk 3 above)
Comment (villejuutila) [reply src/routes/drive-planning.routes.tsx:68]: Resolved

File: src/routes/drive-planning.routes.tsx (hunk 4 above)
Comment (villejuutila) [reply src/routes/drive-planning.routes.tsx:135]: Resolved

File: src/types/index.ts (hunk 5 above)
Comment (villejuutila) [reply src/types/index.ts:20]: Resolved

**Review by lauriLukkarinen — APPROVED** [review]

**Review by lauriLukkarinen — APPROVED** [review]

**Pull request merged**